src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -211,7 +212,20 @@ CreateSingleTaskRouterPlan(Query *originalQuery, Query *query,
 		modifyTask = true;
 	}
 
-	if (modifyTask)
+	if (updateFromQuery)
+	{
+		DeferredErrorMessage *planningError = NULL;
+		planningError = ModifyQuerySupported(query);
+
+		if (planningError != NULL)
+		{
+			multiPlan->planningError = planningError;
+			return multiPlan;
+		}
+
+		task = RouterSelectTask(originalQuery, restrictionContext, &placementList);
```

Comment:
Hey @marcocitus, it turned out to be a bit more complex to send all UPDATE (and maybe DELETE) logic through RouterSelectTask. It seems possible, I did some part in a hacky way on [this branch ](https://github.com/citusdata/citus/commits/update_with_subquery). The problem is that other parts of code are highly coupled with effects of `RouterModifyTask()`, so using `RouterSelectTask()` requires more refactoring, also testing, than my initial thought.

Some of the considerations are;
- Support of prepared statements.
- The error message on unsupported cases - `RouterSelectTask()` needs refactoring to support the previous granularity.
- Evaluation of stable functions on the master.

Given these, I'm inclined to support UPDATE with subqueries as a special case, and use `RouterSelectTask()` just for them. What do you think?